notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+    "This is the ability to join large tables to other large tables without fully loading either one into memory. Both sources involved in the join are scanned in parallel across all workers, the intermediate data for both sources is then redistributed among the workers based on the join column(s) such that rows from both sources with the same values end up in the same worker.\n",
+    "\n",
+    "![](assets/shuffle-join.png)\n",
+    "\n",
+    "In order to use shuffle join the query context must include:\n",
+    "```\n",
+    "{\n",
+    "   \"sqlJoinAlgorithm\":\"sortMerge\"\n",
+    "}\n",
+    "```\n",
+    "\n",
+    "Given that this example is meant to run on the local docker compose deployment, we can't really use two very large tables, so let's try it out with small sources and just pretend they are big. We'll use the `wikipedia` sample data and join it with `wiki_users` profile data. But first we'll need to create the users table because at the time of this writing there wasn't a matching user source handy:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "995aeb98-b5c3-44e3-a4e5-2e13eb3a248c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# since we don't have a source for wiki_users let's\n",
+    "# create one using in-database transformation that\n",
+    "# generates user profiles from the wikipedia data by\n",
+    "# - grouping on \"user\"\n",
+    "# - injecting variability into the \"group\" column using modulus of the __time of the event\n",
+    "# - \"edits\" represent the number of edits done by the user, so just count the number of events\n",
+    "# - calculate the registration time using the minimum __time from events and adjusting it some variable years back in time\n",
+    "# - determine the preferred language of the user based on their earliest channel edit\n",
+    "\n",
+    "sql = '''\n",
+    "REPLACE INTO \"wiki_users\" OVERWRITE ALL\n",
+    "SELECT \n",
+    "  \"user\", \n",
+    "  EARLIEST(\n",
+    "    CASE \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),5) > 3 THEN 'Reviewers' \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),17) > 13 THEN 'Patrollers' \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),23) > 21 THEN 'Bots'\n",
+    "      ELSE 'Autoconfirmed'\n",
+    "    END,\n",
+    "    1024\n",
+    "  ) AS \"group\",\n",
+    "  count(*) \"edits\",\n",
+    "  TIME_SHIFT(MIN(__time), 'P1Y', -1 * MOD(MIN(EXTRACT (MICROSECOND FROM __time)),20) ) AS \"registered_at_ms\",\n",
+    "  EARLIEST(SUBSTRING(\"channel\", 2, 2), 1024) AS \"language\"\n",
+    "FROM \"wikipedia_events\"\n",
+    "GROUP BY 1\n",
+    "PARTITIONED BY ALL\n",
+    "'''\n",
+    "request = druid.sql.sql_request( sql)              # init request object\n",
+    "request.add_context( 'finalizeAggregations', True) # EARLIEST functions will store a partial aggregation otherwise\n",
```

Comment:
partial means that you save the internal aggregation structure rather than the final result. i.e. instead of a just a value it stores the timestamp and the value so that if there are other rows in the result, they can be further aggregated using EARLIEST and still product the correct result.